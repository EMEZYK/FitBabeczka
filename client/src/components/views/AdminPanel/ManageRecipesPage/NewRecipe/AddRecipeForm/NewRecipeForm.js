import React from "react";
import { FieldArray, useFormik, FormikProvider, Field } from "formik";
import axios from "axios";
import {
  Title,
  Form,
  LabelName,
  SaveButton,
  TitleWrapper,
} from "../../../EditUserPage/EditUserProfilePage.styled";
import { TextArea } from "../../../../LoginPage/LoginPage.styled";
import { InputComponent } from "../../../../../ui/Input/Input.styled";
import { DropDownCategory } from "./Dropdown/CategoryDropdown";
import { DifficultyLevelDropdown } from "./Dropdown/DifficultyLevelDropdown";
import { NewRecipeFormSchema } from "./NewRecipeFormSchema";
import { useEffect, useState } from "react";
import { toast, ToastContainer } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import {
  AddIcon,
  AddIngredientsButton,
  IngredientsList,
  OneIngredient as IngredientWrapper,
  IngredientListWrapper,
  ErrorInfo,
  RecipePhoto,
  RecipeImageWrapper,
} from "../../EditMenuPage/EditMenuPage.styled";

import {
  DeleteButton,
  MobileDeleteButton,
  NewRecipeFormWrapper,
  CloseModalButton,
} from "../../EditDeleteModal/Modal.styled";
import { useMediaQuery } from "react-responsive";

export const NewRecipeForm = ({
  setOpenModal,
  modalRecipeTitle,
  addOperation = true,
  recipeId,
  categories,
  imageName: currentImage,
  onEdit,
  setEditedRecipeId,
  setContext,
  // onSuccess,
  // onFail,
}) => {
  const [recipeData, setRecipeData] = useState(null);
  const [newImage, setNewImage] = useState("");
  const [newRecipe, setNewRecipe] = useState(true);

  const formik = useFormik({
    initialValues: {
      name: "",
      description: "",
      ingredients: [],
      preparation: "",
      category: "",
      servingsNumber: "",
      time: "",
      difficultyLevel: "",
      photo: "",
    },
    validationSchema: NewRecipeFormSchema,

    onSubmit: (values) => {
      const formData = new FormData();

      for (let value in values) {
        if (value === "ingredients") {
          formData.append(value, JSON.stringify(values[value]));
          continue;
        }

        formData.append(value, values[value]);
      }

      const getRecipes = () => {
        axios({
          method: "GET",
          url: "/recipes?limit=1000",
        })
          .then((response) => {
            setContext(response.data);
          })
          .catch((err) => {
            console.error(err);
          });
      };

      if (addOperation) {
        axios
          .post("/recipes", formData)
          .then((res) => {
            setOpenModal(false);
            return getRecipes();
          })
          .then((recipesResponse) => {})
          .catch((e) => console.log("Failed to upload recipe", e.message));
      } else {
        axios
          .put(`/recipes/${recipeId}`, formData)
          .then((res) => {
            setEditedRecipeId(recipeId);
            // onSuccess()
            // toast.success("Twój przepis został zaktualizowany");
            setOpenModal(false);

            return getRecipes();
          })
          .then((recipesResponse) => {
            setContext(recipesResponse.data);
          })
          .catch(() => {
            return toast.error(
              "Twój przepis nie został zaktualizowany. Zweryfikuj wprowadzane dane"
            );
          });
      }
    },
  });

  const changeImageSrc = (e) => {
    const srcode = URL.createObjectURL(e.target.files[0]);
    setNewImage(srcode);
    setNewRecipe(false);
  };

  const changeDisplayedImg = () => {
    if (newRecipe === true) {
      return currentImage;
    } else {
      return newImage;
    }
  };

  const autocompleteValues = () => {
    formik.setValues({
      name: recipeData ? recipeData.name : "",
      description: recipeData ? recipeData.description : "",
      ingredients: recipeData ? recipeData.ingredients : [],
      preparation: recipeData ? recipeData.preparation : "",
      category: recipeData ? recipeData.category : "",
      servingsNumber: recipeData ? recipeData.servingsNumber : "",
      time: recipeData ? recipeData.time : "",
      difficultyLevel: recipeData ? recipeData.difficultyLevel : "",
      photo: recipeData ? recipeData.image : "",
    });
  };

  const fetchOneRecipe = async () => {
    await axios({
      url: `/recipes/${recipeId}`,
    })
      .then((response) => {
        setRecipeData(response.data);
      })
      .catch((err) => {
        console.error(err);
      });
  };

  const onCloseClick = () => {
    setOpenModal(false);
  };
  useEffect(() => {
    if (!addOperation) {
      fetchOneRecipe();
      autocompleteValues();
    }
  }, [recipeData?._id]);

  const isMobile = useMediaQuery({
    query: "(max-width: 599px)",
  });

  return (
    <>
      <NewRecipeFormWrapper>
        <TitleWrapper>
          <CloseModalButton
            aria-label="Close"
            type="submit"
            onClick={onCloseClick}
          >
            &times;
          </CloseModalButton>
          <Title role="heading">{modalRecipeTitle}</Title>
        </TitleWrapper>
        <FormikProvider value={formik}>
          <Form
            type="submit"
            onSubmit={formik.handleSubmit}
            encType="multipart/form-data"
          >
            <LabelName htmlFor="dish_name">Nazwa dania</LabelName>
            <InputComponent
              type="text"
              id="name"
              aria-label="name"
              name="name"
              value={formik.values.name}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              required
              border="0.2px solid #DEA8A8"
            />
            {formik.errors.name ? (
              <ErrorInfo>{formik.errors.name}</ErrorInfo>
            ) : null}
            <LabelName htmlFor="description">Krótki opis</LabelName>
            <TextArea
              type="text"
              id="description"
              value={formik.values.description}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              required
              border="0.2px solid #DEA8A8"
            />
            {formik.errors.description ? (
              <ErrorInfo>{formik.errors.description}</ErrorInfo>
            ) : null}
            <LabelName htmlFor="ingredients">Składniki</LabelName>
            <FieldArray name="ingredients">
              {(fieldArrayProps) => {
                const { push, remove, form } = fieldArrayProps;
                const { values } = form;
                const { ingredients } = values;

                return (
                  <>
                    <IngredientListWrapper>
                      <IngredientsList>
                        {values.ingredients && values.ingredients.length > 0
                          ? ingredients.map((value, index) => (
                              <>
                                <IngredientWrapper key={index}>
                                  <Field
                                    name={`ingredients[${index}]`}
                                    type="text"
                                    width="100%"
                                    id={index}
                                  >
                                    {({ field }) => (
                                      <InputComponent
                                        {...field}
                                        type="text"
                                        id="ingredients"
                                        aria-label="ingredients"
                                        placeholder="wpisz składnik"
                                        width="80%"
                                        border="0.2px solid #DEA8A8"
                                        required
                                      />
                                    )}
                                  </Field>

                                  {isMobile === true ? (
                                    <MobileDeleteButton
                                      type="button"
                                      onClick={() => remove(index)}
                                    >
                                      <img
                                        src="/icons/deleteRecipe.svg"
                                        alt=""
                                      />
                                    </MobileDeleteButton>
                                  ) : (
                                    <DeleteButton
                                      type="button"
                                      onClick={() => remove(index)}
                                      padding="0.2rem"
                                      hoverColor="none"
                                      borderRadius="3px"
                                      margin="0 0 0 0.8rem"
                                      height="80%"
                                    >
                                      Usuń
                                    </DeleteButton>
                                  )}
                                </IngredientWrapper>
                              </>
                            ))
                          : ""}

                        <AddIngredientsButton
                          type="button"
                          onClick={() => push()}
                        >
                          <AddIcon
                            src="/icons/addRecipe.svg"
                            mediaWidth="20%"
                          />
                          Dodaj
                        </AddIngredientsButton>
                      </IngredientsList>
                    </IngredientListWrapper>
                  </>
                );
              }}
            </FieldArray>
            {formik.errors.ingredients ? (
              <ErrorInfo>{formik.errors.ingredients}</ErrorInfo>
            ) : null}
            <LabelName htmlFor="preparation">Przygotowanie</LabelName>
            <TextArea
              type="text"
              id="preparation"
              value={formik.values.preparation}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              required
              border="0.2px solid #DEA8A8"
            />
            {formik.errors.preparation ? (
              <ErrorInfo>{formik.errors.preparation}</ErrorInfo>
            ) : null}
            <LabelName htmlFor="category">Kategoria</LabelName>
            <DropDownCategory
              type="text"
              value={formik.values.category}
              onChange={(value) =>
                formik.setFieldValue("category", value.value)
              }
              onBlur={formik.handleBlur}
              required
              categories={categories}
            />
            {formik.errors.category ? (
              <ErrorInfo>{formik.errors.category}</ErrorInfo>
            ) : null}
            <LabelName htmlFor="servings_number">Ilość porcji</LabelName>
            <InputComponent
              type="number"
              min="1"
              aria-label="servings-number"
              id="servingsNumber"
              value={formik.values.servingsNumber}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              required
              border="0.2px solid #DEA8A8"
            />
            {formik.errors.servingsNumber ? (
              <ErrorInfo>{formik.errors.servingsNumber}</ErrorInfo>
            ) : null}
            <LabelName type="text" htmlFor="preparation_time">
              Czas przygotowania
            </LabelName>
            <InputComponent
              type="text"
              id="time"
              aria-label="time"
              value={formik.values.time}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              required
              border="0.2px solid #DEA8A8"
            />
            {formik.errors.time ? (
              <ErrorInfo>{formik.errors.time}</ErrorInfo>
            ) : null}
            <LabelName htmlFor="difficultyLevel">Poziom trudności</LabelName>
            <DifficultyLevelDropdown
              id="difficultyLevel"
              value={formik.values.difficultyLevel}
              onChange={(value) =>
                formik.setFieldValue("difficultyLevel", value.value)
              }
              onBlur={formik.handleBlur}
              required
            />
            {formik.errors.difficultyLevel ? (
              <ErrorInfo>{formik.errors.difficultyLevel}</ErrorInfo>
            ) : null}
            <LabelName htmlFor="photo">Zdjęcie</LabelName>
            <InputComponent
              minHeight="0"
              border="0"
              type="file"
              name="photo"
              aria-label="photo"
              accept="image/*"
              title=" "
              id="img"
              onChange={(e) => {
                changeImageSrc(e);
                formik.setFieldValue("photo", e.currentTarget.files[0]);
              }}
            />
            <RecipeImageWrapper>
              <p>Zdjęcie przepisu:</p>
              <RecipePhoto alt="" src={changeDisplayedImg()} />
            </RecipeImageWrapper>
            {formik.errors.photo ? (
              <ErrorInfo>{formik.errors.photo}</ErrorInfo>
            ) : null}
            <SaveButton type="submit" aria-label="save" onClick={onEdit}>
              Zapisz
            </SaveButton>
          </Form>
        </FormikProvider>
      </NewRecipeFormWrapper>
      <ToastContainer />
    </>
  );
};
