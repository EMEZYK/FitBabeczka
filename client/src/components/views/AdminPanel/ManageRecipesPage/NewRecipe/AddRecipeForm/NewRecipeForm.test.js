import React from "react";
import { render, cleanup } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import { NewRecipeForm } from "./NewRecipeForm";
import "@testing-library/jest-dom";

describe("Test AddRecipeForm components", () => {
  afterEach(cleanup);

  test("Check if takes a snapshot',", () => {
    const { asFragment } = render(<NewRecipeForm />);
    expect(asFragment(<NewRecipeForm />)).toMatchSnapshot();
  });

  test("Check if renders the basic fields", () => {
    render(<NewRecipeForm />);
    expect(screen.getByRole("heading")).toBeInTheDocument();
    expect(screen.getByLabelText("Close")).toBeInTheDocument();
    expect(screen.getByLabelText("name")).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", { name: "Krótki opis" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("textbox", { name: "Przygotowanie" })
    ).toBeInTheDocument();

    expect(screen.getByLabelText("time")).toBeInTheDocument();
    expect(screen.getByLabelText("servings-number")).toBeInTheDocument();
    expect(screen.getByLabelText("photo")).toBeInTheDocument();
    expect(screen.getByLabelText("save")).toBeInTheDocument();
  });

  test("Check if fields are required", () => {
    render(<NewRecipeForm />);
    const nameInput = screen.getByLabelText("name");
    expect(nameInput).toBeRequired();

    const description = screen.getByRole("textbox", { name: "Krótki opis" });
    expect(description).toBeRequired();

    const preparation = screen.getByRole("textbox", { name: "Przygotowanie" });
    expect(preparation).toBeRequired();

    const servingsNumber = screen.getByLabelText("servings-number");
    expect(servingsNumber).toBeRequired();

    const time = screen.getByLabelText("time");
    expect(time).toBeRequired();
  });
});
