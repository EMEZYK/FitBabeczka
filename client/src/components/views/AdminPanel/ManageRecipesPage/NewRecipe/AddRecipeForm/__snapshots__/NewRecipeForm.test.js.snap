// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AddRecipeForm components Check if takes a snapshot', 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ sc-dkzDqf sc-bBrHrO hvLzQT WCzRc hRhRdA"
  >
    <div
      class="sc-iBkjds ozUct"
    >
      <button
        aria-label="Close"
        class="sc-iIPllB lCCfB"
        type="submit"
      >
        ×
      </button>
      <h2
        class="sc-hKMtZM gGDAVZ"
        role="heading"
      />
    </div>
    <form
      class="sc-eCYdqJ brYmts"
      enctype="multipart/form-data"
      type="submit"
    >
      <label
        class="sc-jSMfEi eHcCgu"
        for="dish_name"
      >
        Nazwa dania
      </label>
      <input
        aria-label="name"
        class="sc-crXcEl eBnNGu"
        id="name"
        name="name"
        required=""
        type="text"
        value=""
      />
      <label
        class="sc-jSMfEi eHcCgu"
        for="description"
      >
        Krótki opis
      </label>
      <textarea
        class="sc-kDDrLX bnaXvZ"
        id="description"
        required=""
        type="text"
      />
      <label
        class="sc-jSMfEi eHcCgu"
        for="ingredients"
      >
        Składniki
      </label>
      <div
        class="sc-idiyUo dRBNw"
      >
        <div
          class="sc-fEOsli dDmFoF"
        >
          <button
            class="sc-fnykZs hWZGXu"
            type="button"
          >
            <img
              class="sc-hAZoDl iRxQlp"
              src="/icons/addRecipe.svg"
            />
            Dodaj
          </button>
        </div>
      </div>
      <label
        class="sc-jSMfEi eHcCgu"
        for="preparation"
      >
        Przygotowanie
      </label>
      <textarea
        class="sc-kDDrLX bnaXvZ"
        id="preparation"
        required=""
        type="text"
      />
      <label
        class="sc-jSMfEi eHcCgu"
        for="category"
      >
        Kategoria
      </label>
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-fs4fqg-control"
        >
          <div
            class=" css-319lph-ValueContainer"
          >
            <div
              class=" css-14el2xx-placeholder"
              id="react-select-2-placeholder"
            >
              Select...
            </div>
            <div
              class=" css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          type="hidden"
          value=""
        />
      </div>
      <label
        class="sc-jSMfEi eHcCgu"
        for="servings_number"
      >
        Ilość porcji
      </label>
      <input
        aria-label="servings-number"
        class="sc-crXcEl eBnNGu"
        id="servingsNumber"
        min="1"
        required=""
        type="number"
        value=""
      />
      <label
        class="sc-jSMfEi eHcCgu"
        for="preparation_time"
        type="text"
      >
        Czas przygotowania
      </label>
      <input
        aria-label="time"
        class="sc-crXcEl eBnNGu"
        id="time"
        required=""
        type="text"
        value=""
      />
      <label
        class="sc-jSMfEi eHcCgu"
        for="difficultyLevel"
      >
        Poziom trudności
      </label>
      <div
        class=" css-b62m3t-container"
        id="difficultyLevel"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-fs4fqg-control"
        >
          <div
            class=" css-319lph-ValueContainer"
          >
            <div
              class=" css-14el2xx-placeholder"
              id="react-select-3-placeholder"
            >
              Select...
            </div>
            <div
              class=" css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-3-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-3-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          type="hidden"
          value=""
        />
      </div>
      <label
        class="sc-jSMfEi eHcCgu"
        for="photo"
      >
        Zdjęcie
      </label>
      <input
        accept="image/*"
        aria-label="photo"
        class="sc-crXcEl kVsQAf"
        id="img"
        name="photo"
        title=" "
        type="file"
      />
      <div
        class="sc-hHLeRK gXxFZB"
      >
        <p>
          Zdjęcie przepisu:
        </p>
        <img
          alt=""
          class="sc-dmRaPn dxfbXy"
        />
      </div>
      <button
        aria-label="save"
        class="sc-gsnTZi sc-gKXOVf goHXkf dEJMjr"
        type="submit"
      >
        Zapisz
      </button>
    </form>
  </div>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
